{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in data available through The Arabidopsis Information Resource (TAIR) database\n",
    "The purpose of this notebook is to read in and do a preliminary survey of the data related to text and ontology annotations of text that is available through TAIR. The datasets need to be organized and also restructured into a standard format that will allow it be combined with datasets from other resources. This notebook takes the following input files that were downloaded from the TAIR and produces a set of files that have standard columns, which are listed and described below.\n",
    "\n",
    "### Files read\n",
    "```\n",
    "plant-data/databases/tair/Locus_Germplasm_Phenotype_20190930.txt\n",
    "plant-data/databases/tair/po_anatomy_gene_arabidopsis_tair.assoc\n",
    "plant-data/databases/tair/po_temporal_gene_arabidopsis_tair.assoc\n",
    "plant-data/databases/tair/ATH_GO_GOSLIM.txt\n",
    "```\n",
    "\n",
    "### Files created\n",
    "```\n",
    "plant-data/reshaped_data/tair_phenotype_descriptions.csv\n",
    "plant-data/reshaped_data/tair_all_go_annotations.csv\n",
    "plant-data/reshaped_data/tair_curated_go_annotations.csv\n",
    "plant-data/reshaped_data/tair_curated_po_annotations.csv\n",
    "```\n",
    "\n",
    "### Columns in the created files\n",
    "* **species**: A string indicating what species the gene is in, currently uses the 3-letter codes from the KEGG database.\n",
    "* **unique_gene_identifiers**: Pipe delimited list of gene identifers, names, models, etc which must uniquely refer to this gene.\n",
    "* **other_gene_identifiers**: Pipe delimited list of other identifers, names, aliases, synonyms for the gene, which may but do not have to uniquely refer to it.\n",
    "* **gene_models**: Pipe delimited list of gene model names that map to this gene.\n",
    "* **descriptions**: A free text field for any descriptions of phenotyes associated with this gene.\n",
    "* **annotations**: Pipe delimited list of gene ontology term identifiers.\n",
    "* **sources**: Pipe delimited list of strings that indicate where this data comes from such as database names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize\n",
    "\n",
    "sys.path.append(\"../utils\")\n",
    "from constants import NCBI_TAG, EVIDENCE_CODES, ABBREVIATIONS_MAP\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.nlp.preprocess import concatenate_with_delim, subtract_string_lists, replace_delimiter, concatenate_texts\n",
    "from oats.nlp.small import remove_punctuation, remove_enclosing_brackets, add_prefix_safely\n",
    "\n",
    "OUTPUT_DIR = \"../reshaped_data\"\n",
    "mpl.rcParams[\"figure.dpi\"] = 200\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Columns that should be in the final reshaped files.\n",
    "reshaped_columns = [\"species\", \n",
    " \"unique_gene_identifiers\", \n",
    " \"other_gene_identifiers\", \n",
    " \"gene_models\", \n",
    " \"descriptions\", \n",
    " \"annotations\", \n",
    " \"sources\"]\n",
    "\n",
    "# Creating and testing a lambda for finding gene model strings.\n",
    "gene_model_pattern = re.compile(\"at[0-9]{1}g[0-9]+\")\n",
    "is_gene_model = lambda s: bool(gene_model_pattern.match(s.lower()))\n",
    "assert is_gene_model(\"AT1G34534\") == True\n",
    "assert is_gene_model(\"at2g23452\") == True\n",
    "assert is_gene_model(\"ACAB1\") == False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with genes and phenotype descriptions (Locus_Germplasm_Phenotype_20180702.txt)\n",
    "Reading in the dataset of phenotypic descriptions. There is only one value specified as the gene name (locus name) in the original dataset so this column does not need to be parsed further. The descriptions commmonly use semi-colons to separate phrases. The next cell gets the distribution of the number of phrases in each description field for the dataset of text descriptions, as determined by a sentence parser. The majority of the descriptions are a single sentence or phrase, but some contain more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8769</th>\n",
       "      <td>AT3G63290</td>\n",
       "      <td>reduced male and female gametophytic transmiss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4318</th>\n",
       "      <td>AT2G20890</td>\n",
       "      <td>Increased ROS accumulation in guard cells, inc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9498</th>\n",
       "      <td>AT4G18240</td>\n",
       "      <td>Reduction in starch content. starch turnover  ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14719</th>\n",
       "      <td>EMB1777</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13619</th>\n",
       "      <td>AT5G62690</td>\n",
       "      <td>Left-handed helical growth in the root and oth...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>AT1G05850</td>\n",
       "      <td>The pom1/eli1 double mutant was compromised in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8544</th>\n",
       "      <td>AT3G59420</td>\n",
       "      <td>Increased number of lateral root meristems, gr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7908</th>\n",
       "      <td>AT3G49080</td>\n",
       "      <td>Reduced transmission through male and female g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7073</th>\n",
       "      <td>AT3G19720</td>\n",
       "      <td>Leaf mesophyll cells in mutants contain fewer ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12612</th>\n",
       "      <td>AT5G40280</td>\n",
       "      <td>Enlarged meristems, extra floral organs,retard...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551</th>\n",
       "      <td>AT1G25490</td>\n",
       "      <td>Agravitropic root growth.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1047</th>\n",
       "      <td>AT1G15830</td>\n",
       "      <td>lhcb6 cch double mutant shows ABA insensitivit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11549</th>\n",
       "      <td>AT5G13930</td>\n",
       "      <td>Colorless seed coat.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11768</th>\n",
       "      <td>AT5G17490</td>\n",
       "      <td>The hypocotyl length, in red light and dark gr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1073</th>\n",
       "      <td>AT1G16720</td>\n",
       "      <td>High chlorophyll fluorescence.  Not able to gr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7017</th>\n",
       "      <td>AT3G18730</td>\n",
       "      <td>tsk mutant has reduced root growth, fasciated ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8098</th>\n",
       "      <td>AT3G51160</td>\n",
       "      <td>fucose content in leaf polysaccharides reduced...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9021</th>\n",
       "      <td>AT4G05530</td>\n",
       "      <td>Resistant to (root elongation on) 2,4-dichloro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14471</th>\n",
       "      <td>EMB1489</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3705</th>\n",
       "      <td>AT2G01570</td>\n",
       "      <td>Semi-dwarf phenotype.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      unique_gene_identifiers                                       descriptions\n",
       "8769                AT3G63290  reduced male and female gametophytic transmiss...\n",
       "4318                AT2G20890  Increased ROS accumulation in guard cells, inc...\n",
       "9498                AT4G18240  Reduction in starch content. starch turnover  ...\n",
       "14719                 EMB1777  embryo defective; terminal embryo phenotype: g...\n",
       "13619               AT5G62690  Left-handed helical growth in the root and oth...\n",
       "467                 AT1G05850  The pom1/eli1 double mutant was compromised in...\n",
       "8544                AT3G59420  Increased number of lateral root meristems, gr...\n",
       "7908                AT3G49080  Reduced transmission through male and female g...\n",
       "7073                AT3G19720  Leaf mesophyll cells in mutants contain fewer ...\n",
       "12612               AT5G40280  Enlarged meristems, extra floral organs,retard...\n",
       "1551                AT1G25490                          Agravitropic root growth.\n",
       "1047                AT1G15830  lhcb6 cch double mutant shows ABA insensitivit...\n",
       "11549               AT5G13930                               Colorless seed coat.\n",
       "11768               AT5G17490  The hypocotyl length, in red light and dark gr...\n",
       "1073                AT1G16720  High chlorophyll fluorescence.  Not able to gr...\n",
       "7017                AT3G18730  tsk mutant has reduced root growth, fasciated ...\n",
       "8098                AT3G51160  fucose content in leaf polysaccharides reduced...\n",
       "9021                AT4G05530  Resistant to (root elongation on) 2,4-dichloro...\n",
       "14471                 EMB1489  embryo defective; terminal embryo phenotype: c...\n",
       "3705                AT2G01570                              Semi-dwarf phenotype."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"../databases/tair/Locus_Germplasm_Phenotype_20190930.txt\"\n",
    "usecols = [\"LOCUS_NAME\", \"PHENOTYPE\"]\n",
    "usenames = [\"unique_gene_identifiers\", \"descriptions\"]\n",
    "renamed = {k:v for k,v in zip(usecols,usenames)}\n",
    "df = pd.read_table(filename, usecols=usecols)\n",
    "df.rename(columns=renamed, inplace=True)\n",
    "df.dropna(axis=\"rows\",inplace=True)\n",
    "df.sample(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15442, 2)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15442, 2)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting distributions of number of phrases in each description.\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "ax1.set_title(\"Phenotype Descriptions\")\n",
    "ax2.set_title(\"Phenotype Descriptions\")\n",
    "ax1.set_xlabel(\"Number of phrases\")\n",
    "ax2.set_xlabel(\"Number of words\")\n",
    "x1 = [len(sent_tokenize(x)) for x in df[\"descriptions\"].values]\n",
    "x2 = [len(word_tokenize(x)) for x in df[\"descriptions\"].values]\n",
    "ax1.hist(x1, bins=15, range=(0,15), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "ax2.hist(x2, bins=30, range=(0,200), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "fig.set_size_inches(15,4)\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ACL2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>flower stems are much reduced in length, semi-...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>ADE1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sustained and enhanced levels of ABA-regulated...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALB2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>white embryo and seedling (albino), lethal.</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-104</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-128</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>increased aluminum sensitivity; poor root grow...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum sens...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mesophyll cells contain approximately half the...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>fewer and larger chloroplasts than wild type i...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>pale first leaves in early seedling growth; me...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC8</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect;  mea...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARS1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Semi-dominant arsenic resistance phenotype. In...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers other_gene_identifiers gene_models                                       descriptions annotations sources\n",
       "0      ath                    ACL2                                     flower stems are much reduced in length, semi-...                TAIR\n",
       "1      ath                    ADE1                                     sustained and enhanced levels of ABA-regulated...                TAIR\n",
       "2      ath                    ALB2                                           white embryo and seedling (albino), lethal.                TAIR\n",
       "3      ath                 ALR-104                                     incomplete penetrance; increased aluminum resi...                TAIR\n",
       "4      ath                 ALR-128                                     incomplete penetrance; increased aluminum resi...                TAIR\n",
       "5      ath                    ALS4                                     increased aluminum sensitivity; poor root grow...                TAIR\n",
       "6      ath                    ALS5                                     incomplete penetrance; increased aluminum sens...                TAIR\n",
       "7      ath                   ARC10                                     no visible whole plant phenotypic effect; mean...                TAIR\n",
       "8      ath                    ARC2                                     mesophyll cells contain approximately half the...                TAIR\n",
       "9      ath                    ARC4                                     fewer and larger chloroplasts than wild type i...                TAIR\n",
       "10     ath                    ARC7                                     pale first leaves in early seedling growth; me...                TAIR\n",
       "11     ath                    ARC8                                     no visible whole plant phenotypic effect;  mea...                TAIR\n",
       "12     ath                    ARC9                                     no visible whole plant phenotypic effect; mean...                TAIR\n",
       "13     ath                    ARS1                                     Semi-dominant arsenic resistance phenotype. In...                TAIR\n",
       "14     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "15     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "16     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "17     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "18     ath               AT1G01040                          AT1G01040                         embryo defective; globular                TAIR\n",
       "19     ath               AT1G01040                          AT1G01040                         embryo defective; globular                TAIR"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all expected column names.\n",
    "df[\"species\"] = \"ath\"\n",
    "df[\"other_gene_identifiers\"] = \"\"\n",
    "df[\"gene_models\"] = df[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df[\"annotations\"] = \"\"\n",
    "df[\"sources\"] = \"TAIR\"\n",
    "df = df[reshaped_columns]\n",
    "\n",
    "# Outputting the dataset of phenotype descriptions to csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_phenotype_descriptions.csv\")\n",
    "df.to_csv(path, index=False)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with gene ontology annotations (ATH_GO_GOSLIM.txt)\n",
    "Read in the file containing names of loci and corresponding information relating to gene ontology term annotation. Not all of the columns are used here, only a subset of them are read in. The relationship column refers to the relationships between the gene for that loci and the term mentioned on that given line. Evidence refer to the method of acquiring and the confidence in the annotation itself. This is retained so that we can subset that dataset based on whether the annotations are experimentally confirmed or simply predicted annotations. This section also looks at how many unique values are present for each field.\n",
    "\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. About half of the term annotations were made computationally, but there are also a high number of annotations available from high confidence annotations such as experimentally validated, curator statements, and author statements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       locus       object     term_id evidence_code                              reference\n",
      "0  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "1  AT1G01010    AT1G01010  GO:0006355           ISS    Publication:1345963|PMID:11118137  \n",
      "2  AT1G01010    AT1G01010  GO:0044212           IPI  Publication:501786139|PMID:30356219  \n",
      "3  AT1G01010    AT1G01010  GO:0005634           IEA            AnalysisReference:501756971\n",
      "4  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "5  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "6  AT1G01010    AT1G01010  GO:0006355           IEA            AnalysisReference:501756966\n",
      "7  AT1G01010  AT1G01010.1  GO:0005634           ISM            AnalysisReference:501780126\n",
      "8  AT1G01010    AT1G01010  GO:0006355           ISS    Publication:1345963|PMID:11118137  \n",
      "9  AT1G01010    AT1G01010  GO:0003700           ISS    Publication:1345963|PMID:11118137  \n",
      "(343166, 7)\n",
      "locus                31302\n",
      "object               68581\n",
      "relationship            19\n",
      "term_label            7317\n",
      "term_id               7317\n",
      "evidence_code           16\n",
      "reference            12158\n"
     ]
    }
   ],
   "source": [
    "filename = \"../databases/tair/ATH_GO_GOSLIM.txt\"\n",
    "df_go = pd.read_table(filename, header=None, usecols=[0,2,3,4,5,9,12])\n",
    "df_go.columns = [\"locus\",\"object\",\"relationship\",\"term_label\",\"term_id\",\"evidence_code\",\"reference\"]\n",
    "unique_values = {col:len(pd.unique(df_go[col].values)) for col in df_go.columns}\n",
    "print(df_go[[\"locus\",\"object\",\"term_id\",\"evidence_code\",\"reference\"]].head(10))\n",
    "print(df_go.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "high_throughput              5914\n",
      "computational               47057\n",
      "curator_statement           22809\n",
      "experimental               131557\n",
      "electronic_annotation       58110\n",
      "author_statement            17092\n",
      "phylogenetics               60627\n"
     ]
    }
   ],
   "source": [
    "code_quantities = {c:len([x for x in df_go[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0044212</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0003677</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016021</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0044212</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0032541</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers other_gene_identifiers gene_models descriptions annotations sources\n",
       "0      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "1      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "2      ath               AT1G01010                          AT1G01010               GO:0044212    TAIR\n",
       "3      ath               AT1G01010                          AT1G01010               GO:0005634    TAIR\n",
       "4      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "5      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "6      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "7      ath               AT1G01010                          AT1G01010               GO:0005634    TAIR\n",
       "8      ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "9      ath               AT1G01010                          AT1G01010               GO:0003700    TAIR\n",
       "10     ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "11     ath               AT1G01010                          AT1G01010               GO:0003677    TAIR\n",
       "12     ath               AT1G01010                          AT1G01010               GO:0016021    TAIR\n",
       "13     ath               AT1G01010                          AT1G01010               GO:0044212    TAIR\n",
       "14     ath               AT1G01010                          AT1G01010               GO:0006355    TAIR\n",
       "15     ath               AT1G01020                          AT1G01020               GO:0016125    TAIR\n",
       "16     ath               AT1G01020                          AT1G01020               GO:0032541    TAIR\n",
       "17     ath               AT1G01020                          AT1G01020               GO:0016125    TAIR\n",
       "18     ath               AT1G01020                          AT1G01020               GO:0005739    TAIR\n",
       "19     ath               AT1G01020                          AT1G01020               GO:0016125    TAIR"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df_go[\"species\"] = \"ath\"\n",
    "df_go[\"unique_gene_identifiers\"] = df_go[\"locus\"]\n",
    "df_go[\"other_gene_identifiers\"] = \"\"\n",
    "df_go[\"gene_models\"] = df_go[\"unique_gene_identifiers\"].map(lambda x: \"\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_go[\"descriptions\"] = \"\"\n",
    "df_go[\"annotations\"] = df_go[\"term_id\"]\n",
    "df_go[\"sources\"] = \"TAIR\"\n",
    "high_confidence_categories = [\"experimental\",\"author_statement\",\"curator_statement\"]\n",
    "df_go[\"high_confidence\"] = df_go[\"evidence_code\"].apply(lambda x: EVIDENCE_CODES[x] in high_confidence_categories)\n",
    "\n",
    "# Subset to only include the high-quality GO annotations and ouptut the dataset to a csv file.\n",
    "df_go_high_confidence = df_go[df_go[\"high_confidence\"]==True]\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_curated_go_annotations.csv\")\n",
    "df_go_high_confidence = df_go_high_confidence[reshaped_columns]\n",
    "df_go_high_confidence.to_csv(path, index=False)\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "df_go = df_go[reshaped_columns]\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_all_go_annotations.csv\")\n",
    "df_go.to_csv(path, index=False)\n",
    "df_go.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with plant ontology term annotations (po_[termporal/anatomy]_gene_arabidopsis_tair.assoc)\n",
    "There are two separate files available that include annotations of PO terms. The files do not have headers so column names are added based on how the columns are described in the accompanying available readme files. One of the files contains annotations for PO terms that are spatial, or describe a specific part of plant anatomy or plant molecular structures. The other file contains annotations for PO terms that are temporal, or refer to a specific process or stage of development. These files are each read in separately, and the next cells look at the quantity of unique values in the columns of each dataset. There are more spatial annotations than temporal annotations, and a greater number of terms used to describe the spatial annotations.\n",
    "\n",
    "The next field combines the two datasets of PO annotations and looks at the number of unique values for each column in the resulting dataset. Because there is no overlap in the terms between the two, the datasets are simply appended to one another and the total unique terms are a sum of the individual datasets.\n",
    "\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. Almost all of the PO term annotations are high confidence, they are experimentally validated, and only a few of them are derived from author statements.\n",
    "\n",
    "The strings which are described in the synonyms column are included as references to each gene, and are combined with the gene name mentioned in the symbol column into a single bar delimited list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(371438, 7)\n",
      "symbol               24085\n",
      "term_id                369\n",
      "references            2730\n",
      "evidence_code            8\n",
      "name                 25422\n",
      "synonyms             28093\n",
      "type                     6\n"
     ]
    }
   ],
   "source": [
    "# Reading in the dataset of spatial PO term annotations.\n",
    "filename = \"../databases/tair/po_anatomy_gene_arabidopsis_tair.assoc\"\n",
    "df_po_spatial = pd.read_table(filename, header=None, skiprows=0, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_spatial.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_spatial[col].values)) for col in df_po_spatial.columns}\n",
    "print(df_po_spatial.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(186819, 7)\n",
      "symbol               18006\n",
      "term_id                 91\n",
      "references             478\n",
      "evidence_code            4\n",
      "name                 18523\n",
      "synonyms             18973\n",
      "type                     5\n"
     ]
    }
   ],
   "source": [
    "# Reading in the dataset of temporal PO term annotations.\n",
    "filename = \"../databases/tair/po_temporal_gene_arabidopsis_tair.assoc\"\n",
    "df_po_temporal = pd.read_table(filename, header=None, skiprows=0, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_temporal.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_temporal[col].values)) for col in df_po_temporal.columns}\n",
    "print(df_po_temporal.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   symbol                                           synonyms evidence_code\n",
      "0  ATARCA  AT1G18080|ATARCA|RACK1A_AT|RACK1A|SAC53|AtRACK...           IEP\n",
      "1    HMG1  AT1G76490|HMG1|HMGR1|AtHMGR1|MAD3|hydroxy meth...           TAS\n",
      "2    BAS1  AT2G26710|BAS1|CYP734A1|CYP72B1|PHYB ACTIVATIO...           TAS\n",
      "3    BAK1  AT4G33430|BAK1|RKS10|SERK3|ELG|ATSERK3|ATBAK1|...           TAS\n",
      "4    BRI1  AT4G39400|BRI1|CBB2|DWF2|BIN1|ATBRI1|BRASSINOS...           TAS\n",
      "5    ATJ6    AT5G06910|ATJ6|J-domain protein 6|MOJ9.8|MOJ9_8           IEP\n",
      "6    XPB2  AT5G41360|XPB2|ATXPB2|homolog of Xeroderma pig...           NAS\n",
      "7    XPB1  AT5G41370|XPB1|ATXPB1|homolog of xeroderma pig...           NAS\n",
      "8   PPCK2  AT3G04530|PPCK2|PEPCK2|ATPPCK2|phosphoenolpyru...           IEP\n",
      "9    MYC2  AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...           NAS\n",
      "(558257, 7)\n",
      "symbol               24148\n",
      "term_id                460\n",
      "references            2807\n",
      "evidence_code            8\n",
      "name                 25506\n",
      "synonyms             28238\n",
      "type                     7\n"
     ]
    }
   ],
   "source": [
    "# Looking at how many unique values each column has.\n",
    "df_po = df_po_spatial.append(df_po_temporal, ignore_index=True)\n",
    "unique_values = {col:len(pd.unique(df_po[col].values)) for col in df_po.columns}\n",
    "print(df_po[[\"symbol\",\"synonyms\",\"evidence_code\"]].head(10))\n",
    "print(df_po.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "high_throughput              4527\n",
      "computational                   0\n",
      "curator_statement               0\n",
      "experimental               553612\n",
      "electronic_annotation           0\n",
      "author_statement              118\n",
      "phylogenetics                   0\n"
     ]
    }
   ],
   "source": [
    "# Quantifying the number of annotations of each type.\n",
    "code_quantities = {c:len([x for x in df_po[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol           object\n",
       "term_id          object\n",
       "references       object\n",
       "evidence_code    object\n",
       "name             object\n",
       "synonyms         object\n",
       "type             object\n",
       "dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_po['name'] = df_po['name'].astype(str)\n",
    "df_po.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-96b1355f1ac4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"annotations\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"term_id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"sources\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"TAIR\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"other_gene_identifiers\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0msubtract_string_lists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"|\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"other_gene_identifiers\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"unique_gene_identifiers\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0mdf_po\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mreshaped_columns\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, axis, raw, result_type, args, **kwds)\u001b[0m\n\u001b[1;32m   6876\u001b[0m             \u001b[0mkwds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6877\u001b[0m         )\n\u001b[0;32m-> 6878\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6879\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6880\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapplymap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;34m\"DataFrame\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mget_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    184\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    185\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 186\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    187\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    188\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mapply_empty_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    294\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    295\u001b[0m                 result = libreduction.compute_reduction(\n\u001b[0;32m--> 296\u001b[0;31m                     \u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdummy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdummy\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    297\u001b[0m                 )\n\u001b[1;32m    298\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mValueError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/reduction.pyx\u001b[0m in \u001b[0;36mpandas._libs.reduction.compute_reduction\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/reduction.pyx\u001b[0m in \u001b[0;36mpandas._libs.reduction.Reducer.get_result\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m<ipython-input-21-96b1355f1ac4>\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"annotations\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"term_id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"sources\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"TAIR\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"other_gene_identifiers\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0msubtract_string_lists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"|\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"other_gene_identifiers\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"unique_gene_identifiers\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0mdf_po\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_po\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mreshaped_columns\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    869\u001b[0m         \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    870\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 871\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    872\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    873\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_value\u001b[0;34m(self, series, key)\u001b[0m\n\u001b[1;32m   4398\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mInvalidIndexError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4399\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4400\u001b[0;31m         \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues_from_object\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4401\u001b[0m         \u001b[0mk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues_from_object\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4402\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.values_from_object\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m_internal_get_values\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    528\u001b[0m         \"\"\"\n\u001b[1;32m    529\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 530\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    531\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    532\u001b[0m     \u001b[0;31m# ops\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mget_values\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1568\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1569\u001b[0m         \u001b[0;34m\"\"\" return a dense type view \"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1570\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_block\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_dense\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1571\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1572\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/envs/oats/lib/python3.6/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m_block\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1520\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1521\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_block\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1522\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1523\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1524\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "combine_columns = lambda row, columns: concatenate_with_delim(\"|\", [row[column] for column in columns])\n",
    "df_po[\"species\"] = \"ath\"\n",
    "df_po[\"unique_gene_identifiers\"] = df_po.apply(lambda x: combine_columns(x, [\"symbol\", \"name\"]), axis=1)\n",
    "df_po[\"other_gene_identifiers\"] = df_po[\"synonyms\"]\n",
    "df_po[\"gene_model_strings_1\"] =  df_po[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_po[\"gene_model_strings_2\"] =  df_po[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_po[\"gene_models\"] = df_po.apply(lambda x: combine_columns(x, [\"gene_model_strings_1\", \"gene_model_strings_2\"]), axis=1)\n",
    "df_po[\"descriptions\"] = \"\"\n",
    "df_po[\"annotations\"] = df_po[\"term_id\"]\n",
    "df_po[\"sources\"] = \"TAIR\"\n",
    "df_po[\"other_gene_identifiers\"] = df_po.apply(lambda row: subtract_string_lists(\"|\", row[\"other_gene_identifiers\"],row[\"unique_gene_identifiers\"]), axis=1)\n",
    "df_po = df_po[reshaped_columns]\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_curated_po_annotations.csv\")\n",
    "df_po.to_csv(path, index=False)\n",
    "df_po.head(30)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
