{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in data available through The Arabidopsis Information Resource (TAIR) database\n",
    "The purpose of this notebook is to read in and do a preliminary survey of the data related to text and ontology annotations of text that is available through TAIR. The datasets need to be organized and also restructured into a standard format that will allow it be combined with datasets from other resources. This notebook takes the following input files that were downloaded from the TAIR and produces a set of files that have standard columns, which are listed and described below.\n",
    "\n",
    "### Files read\n",
    "```\n",
    "plant-data/databases/tair/Locus_Germplasm_Phenotype_20180702.txt\n",
    "plant-data/databases/tair/po_anatomy_gene_arabidopsis_tair.assoc\n",
    "plant-data/databases/tair/po_temporal_gene_arabidopsis_tair.assoc\n",
    "plant-data/databases/tair/ATH_GO_GOSLIM.txt\n",
    "```\n",
    "\n",
    "### Files created\n",
    "```\n",
    "plant-data/reshaped_data/tair_phenotype_descriptions.csv\n",
    "plant-data/reshaped_data/tair_all_go_annotations.csv\n",
    "plant-data/reshaped_data/tair_curated_go_annotations.csv\n",
    "plant-data/reshaped_data/tair_curated_po_annotations.csv\n",
    "```\n",
    "\n",
    "### Columns in the created files\n",
    "* **species**: A string indicating what species the gene is in, currently uses the 3-letter codes from the KEGG database.\n",
    "* **unique_gene_identifiers**: Pipe delimited list of gene identifers, names, models, etc which must uniquely refer to this gene.\n",
    "* **other_gene_identifiers**: Pipe delimited list of other identifers, names, aliases, synonyms for the gene, which may but do not have to uniquely refer to it.\n",
    "* **gene_models**: Pipe delimited list of gene model names that map to this gene.\n",
    "* **descriptions**: A free text field for any descriptions of phenotyes associated with this gene.\n",
    "* **annotations**: Pipe delimited list of gene ontology term identifiers.\n",
    "* **sources**: Pipe delimited list of strings that indicate where this data comes from such as database names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import re\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from nltk.tokenize import word_tokenize\n",
    "from nltk.tokenize import sent_tokenize\n",
    "\n",
    "sys.path.append(\"../utils\")\n",
    "from constants import NCBI_TAG, EVIDENCE_CODES, ABBREVIATIONS_MAP\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.nlp.preprocess import concatenate_with_delim, subtract_string_lists, replace_delimiter, concatenate_texts\n",
    "from oats.nlp.small import remove_punctuation, remove_enclosing_brackets, add_prefix_safely\n",
    "\n",
    "OUTPUT_DIR = \"../reshaped_data\"\n",
    "mpl.rcParams[\"figure.dpi\"] = 200\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Columns that should be in the final reshaped files.\n",
    "reshaped_columns = [\"species\", \n",
    " \"unique_gene_identifiers\", \n",
    " \"other_gene_identifiers\", \n",
    " \"gene_models\", \n",
    " \"descriptions\", \n",
    " \"annotations\", \n",
    " \"sources\"]\n",
    "\n",
    "# Creating and testing a lambda for finding gene model strings.\n",
    "gene_model_pattern = re.compile(\"at[0-9]{1}g[0-9]+\")\n",
    "is_gene_model = lambda s: bool(gene_model_pattern.match(s.lower()))\n",
    "assert is_gene_model(\"AT1G34534\") == True\n",
    "assert is_gene_model(\"at2g23452\") == True\n",
    "assert is_gene_model(\"ACAB1\") == False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with genes and phenotype descriptions (Locus_Germplasm_Phenotype_20180702.txt)\n",
    "Reading in the dataset of phenotypic descriptions. There is only one value specified as the gene name (locus name) in the original dataset so this column does not need to be parsed further. The descriptions commmonly use semi-colons to separate phrases. The next cell gets the distribution of the number of phrases in each description field for the dataset of text descriptions, as determined by a sentence parser. The majority of the descriptions are a single sentence or phrase, but some contain more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4455</th>\n",
       "      <td>AT2G33860</td>\n",
       "      <td>very strong allele;homozygotes have flowers wi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1253</th>\n",
       "      <td>AT1G23310</td>\n",
       "      <td>Mutant plants had a pale green color and grew ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10469</th>\n",
       "      <td>AT5G23570</td>\n",
       "      <td>Infection with cucumber mosaic virus (CMV) rev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1764</th>\n",
       "      <td>AT1G43850</td>\n",
       "      <td>Resembles seu-1 single mutant with reduced sta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10722</th>\n",
       "      <td>AT5G35750</td>\n",
       "      <td>Early germination compared to wild-type.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4866</th>\n",
       "      <td>AT2G40320</td>\n",
       "      <td>No obvious visible phenotype.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11819</th>\n",
       "      <td>AT5G61850</td>\n",
       "      <td>intermediate/strong allele; early-arising flow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8875</th>\n",
       "      <td>AT4G32910</td>\n",
       "      <td>Homozygotes are hypersensitivite to ABA and Na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3212</th>\n",
       "      <td>AT2G01570</td>\n",
       "      <td>Much reduced fertility compared to wild type Ler.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11450</th>\n",
       "      <td>AT5G53760</td>\n",
       "      <td>Aberrant root thigmomorphogenesis;  anisotropi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1971</th>\n",
       "      <td>AT1G52340</td>\n",
       "      <td>After bolting, the wilty phenotype became evid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5128</th>\n",
       "      <td>AT2G45350</td>\n",
       "      <td>Plants have impaired NDH (chloroplast Na(P)DH ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9062</th>\n",
       "      <td>AT4G35920</td>\n",
       "      <td>primary roots failed to penetrate a harder aga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665</th>\n",
       "      <td>AT3G43190</td>\n",
       "      <td>The amounts of glucose, fructose, sucrose, cel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5510</th>\n",
       "      <td>AT3G05540</td>\n",
       "      <td>Homozygous SALK_045146 insertional mutants dis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12353</th>\n",
       "      <td>EMB1140</td>\n",
       "      <td>embryo defective; terminal embryo phenotype: c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11176</th>\n",
       "      <td>AT5G48100</td>\n",
       "      <td>Seeds from the homozygous mutant plants were g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9668</th>\n",
       "      <td>AT5G07990</td>\n",
       "      <td>pale brown seeds due to reduced levels of brow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9520</th>\n",
       "      <td>AT5G04160</td>\n",
       "      <td>kanamycin resistant, no visible phenotype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8890</th>\n",
       "      <td>AT4G33000</td>\n",
       "      <td>hypersensitive to NaCl.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      unique_gene_identifiers                                       descriptions\n",
       "4455                AT2G33860  very strong allele;homozygotes have flowers wi...\n",
       "1253                AT1G23310  Mutant plants had a pale green color and grew ...\n",
       "10469               AT5G23570  Infection with cucumber mosaic virus (CMV) rev...\n",
       "1764                AT1G43850  Resembles seu-1 single mutant with reduced sta...\n",
       "10722               AT5G35750           Early germination compared to wild-type.\n",
       "4866                AT2G40320                      No obvious visible phenotype.\n",
       "11819               AT5G61850  intermediate/strong allele; early-arising flow...\n",
       "8875                AT4G32910  Homozygotes are hypersensitivite to ABA and Na...\n",
       "3212                AT2G01570  Much reduced fertility compared to wild type Ler.\n",
       "11450               AT5G53760  Aberrant root thigmomorphogenesis;  anisotropi...\n",
       "1971                AT1G52340  After bolting, the wilty phenotype became evid...\n",
       "5128                AT2G45350  Plants have impaired NDH (chloroplast Na(P)DH ...\n",
       "9062                AT4G35920  primary roots failed to penetrate a harder aga...\n",
       "6665                AT3G43190  The amounts of glucose, fructose, sucrose, cel...\n",
       "5510                AT3G05540  Homozygous SALK_045146 insertional mutants dis...\n",
       "12353                 EMB1140  embryo defective; terminal embryo phenotype: c...\n",
       "11176               AT5G48100  Seeds from the homozygous mutant plants were g...\n",
       "9668                AT5G07990  pale brown seeds due to reduced levels of brow...\n",
       "9520                AT5G04160          kanamycin resistant, no visible phenotype\n",
       "8890                AT4G33000                            hypersensitive to NaCl."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"../databases/tair/Locus_Germplasm_Phenotype_20180702.txt\"\n",
    "usecols = [\"LOCUS_NAME\", \"PHENOTYPE\"]\n",
    "usenames = [\"unique_gene_identifiers\", \"descriptions\"]\n",
    "renamed = {k:v for k,v in zip(usecols,usenames)}\n",
    "df = pd.read_table(filename, usecols=usecols)\n",
    "df.rename(columns=renamed, inplace=True)\n",
    "df.sample(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting distributions of number of phrases in each description.\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "ax1.set_title(\"Phenotype Descriptions\")\n",
    "ax2.set_title(\"Phenotype Descriptions\")\n",
    "ax1.set_xlabel(\"Number of phrases\")\n",
    "ax2.set_xlabel(\"Number of words\")\n",
    "x1 = [len(sent_tokenize(x)) for x in df[\"descriptions\"].values]\n",
    "x2 = [len(word_tokenize(x)) for x in df[\"descriptions\"].values]\n",
    "ax1.hist(x1, bins=15, range=(0,15), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "ax2.hist(x2, bins=30, range=(0,200), density=False, alpha=0.8, histtype='stepfilled', color=\"black\", edgecolor='none')\n",
    "fig.set_size_inches(15,4)\n",
    "fig.tight_layout()\n",
    "fig.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ACL2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>flower stems are much reduced in length, semi-...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>ADE1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>sustained and enhanced levels of ABA-regulated...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALB2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>white embryo and seedling (albino), lethal.</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-104</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALR-128</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum resi...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>increased aluminum sensitivity; poor root grow...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>ALS5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>incomplete penetrance; increased aluminum sens...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC10</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>mesophyll cells contain approximately half the...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>fewer and larger chloroplasts than wild type i...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC7</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>pale first leaves in early seedling growth; me...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC8</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect;  mea...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARC9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>no visible whole plant phenotypic effect; mean...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>ARS1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Semi-dominant arsenic resistance phenotype. In...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; developmental arrest of muta...</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>embryo defective; globular</td>\n",
       "      <td></td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers other_gene_identifiers gene_models                                       descriptions annotations sources\n",
       "0      ath                    ACL2                                     flower stems are much reduced in length, semi-...                TAIR\n",
       "1      ath                    ADE1                                     sustained and enhanced levels of ABA-regulated...                TAIR\n",
       "2      ath                    ALB2                                           white embryo and seedling (albino), lethal.                TAIR\n",
       "3      ath                 ALR-104                                     incomplete penetrance; increased aluminum resi...                TAIR\n",
       "4      ath                 ALR-128                                     incomplete penetrance; increased aluminum resi...                TAIR\n",
       "5      ath                    ALS4                                     increased aluminum sensitivity; poor root grow...                TAIR\n",
       "6      ath                    ALS5                                     incomplete penetrance; increased aluminum sens...                TAIR\n",
       "7      ath                   ARC10                                     no visible whole plant phenotypic effect; mean...                TAIR\n",
       "8      ath                    ARC2                                     mesophyll cells contain approximately half the...                TAIR\n",
       "9      ath                    ARC4                                     fewer and larger chloroplasts than wild type i...                TAIR\n",
       "10     ath                    ARC7                                     pale first leaves in early seedling growth; me...                TAIR\n",
       "11     ath                    ARC8                                     no visible whole plant phenotypic effect;  mea...                TAIR\n",
       "12     ath                    ARC9                                     no visible whole plant phenotypic effect; mean...                TAIR\n",
       "13     ath                    ARS1                                     Semi-dominant arsenic resistance phenotype. In...                TAIR\n",
       "14     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "15     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "16     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "17     ath               AT1G01040                          AT1G01040  embryo defective; developmental arrest of muta...                TAIR\n",
       "18     ath               AT1G01040                          AT1G01040                         embryo defective; globular                TAIR\n",
       "19     ath               AT1G01040                          AT1G01040                         embryo defective; globular                TAIR"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all expected column names.\n",
    "df[\"species\"] = \"ath\"\n",
    "df[\"other_gene_identifiers\"] = \"\"\n",
    "df[\"gene_models\"] = df[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df[\"annotations\"] = \"\"\n",
    "df[\"sources\"] = \"TAIR\"\n",
    "df = df[reshaped_columns]\n",
    "\n",
    "# Outputting the dataset of phenotype descriptions to csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_phenotype_descriptions.csv\")\n",
    "df.to_csv(path, index=False)\n",
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with gene ontology annotations (ATH_GO_GOSLIM.txt)\n",
    "Read in the file containing names of loci and corresponding information relating to gene ontology term annotation. Not all of the columns are used here, only a subset of them are read in. The relationship column refers to the relationships between the gene for that loci and the term mentioned on that given line. Evidence refer to the method of acquiring and the confidence in the annotation itself. This is retained so that we can subset that dataset based on whether the annotations are experimentally confirmed or simply predicted annotations. This section also looks at how many unique values are present for each field.\n",
    "\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. About half of the term annotations were made computationally, but there are also a high number of annotations available from high confidence annotations such as experimentally validated, curator statements, and author statements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       locus       object     term_id evidence_code                    reference\n",
      "0  AT1G01010    AT1G01010  GO:0007275           ISS      Communication:501714663\n",
      "1  AT1G01010  AT1G01010.1  GO:0005634           ISM  AnalysisReference:501750651\n",
      "2  AT1G01010    AT1G01010  GO:0003700           ISS          Publication:1345963\n",
      "3  AT1G01010    AT1G01010  GO:0043090           RCA        Publication:501748728\n",
      "4  AT1G01010    AT1G01010  GO:0006888           RCA        Publication:501748728\n",
      "5  AT1G01020    AT1G01020  GO:0006665           IMP        Publication:501719026\n",
      "6  AT1G01020    AT1G01020  GO:0016125           IMP        Publication:501719026\n",
      "7  AT1G01020    AT1G01020  GO:0005783           IDA        Publication:501719026\n",
      "8  AT1G01020    AT1G01020  GO:0016020           ISS      Communication:501714663\n",
      "9  AT1G01020  AT1G01020.2  GO:0005739           ISM  AnalysisReference:501750651\n",
      "(251297, 7)\n",
      "locus                30251\n",
      "object               57190\n",
      "relationship            19\n",
      "term_label            4924\n",
      "term_id               4924\n",
      "evidence_code           13\n",
      "reference             9149\n"
     ]
    }
   ],
   "source": [
    "filename = \"../databases/tair/ATH_GO_GOSLIM.txt\"\n",
    "df_go = pd.read_table(filename, header=None, usecols=[0,2,3,4,5,9,12])\n",
    "df_go.columns = [\"locus\",\"object\",\"relationship\",\"term_label\",\"term_id\",\"evidence_code\",\"reference\"]\n",
    "unique_values = {col:len(pd.unique(df_go[col].values)) for col in df_go.columns}\n",
    "print(df_go[[\"locus\",\"object\",\"term_id\",\"evidence_code\",\"reference\"]].head(10))\n",
    "print(df_go.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "experimental                85701\n",
      "computational              125915\n",
      "phylogenetics                1671\n",
      "high_throughput                 0\n",
      "author_statement            13460\n",
      "curator_statement           24550\n",
      "electronic_annotation           0\n"
     ]
    }
   ],
   "source": [
    "code_quantities = {c:len([x for x in df_go[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0007275</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0043090</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006888</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006665</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005783</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0016125</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006665</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>GO:0048366</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td></td>\n",
       "      <td>GO:0009908</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers other_gene_identifiers gene_models descriptions annotations sources\n",
       "0      ath               AT1G01010                          AT1G01010               GO:0007275    TAIR\n",
       "1      ath               AT1G01010                          AT1G01010               GO:0005634    TAIR\n",
       "2      ath               AT1G01010                          AT1G01010               GO:0003700    TAIR\n",
       "3      ath               AT1G01010                          AT1G01010               GO:0043090    TAIR\n",
       "4      ath               AT1G01010                          AT1G01010               GO:0006888    TAIR\n",
       "5      ath               AT1G01020                          AT1G01020               GO:0006665    TAIR\n",
       "6      ath               AT1G01020                          AT1G01020               GO:0016125    TAIR\n",
       "7      ath               AT1G01020                          AT1G01020               GO:0005783    TAIR\n",
       "8      ath               AT1G01020                          AT1G01020               GO:0016020    TAIR\n",
       "9      ath               AT1G01020                          AT1G01020               GO:0005739    TAIR\n",
       "10     ath               AT1G01020                          AT1G01020               GO:0005739    TAIR\n",
       "11     ath               AT1G01020                          AT1G01020               GO:0016125    TAIR\n",
       "12     ath               AT1G01020                          AT1G01020               GO:0003674    TAIR\n",
       "13     ath               AT1G01020                          AT1G01020               GO:0006665    TAIR\n",
       "14     ath               AT1G01030                          AT1G01030               GO:0006355    TAIR\n",
       "15     ath               AT1G01030                          AT1G01030               GO:0003700    TAIR\n",
       "16     ath               AT1G01030                          AT1G01030               GO:0048366    TAIR\n",
       "17     ath               AT1G01030                          AT1G01030               GO:0006355    TAIR\n",
       "18     ath               AT1G01030                          AT1G01030               GO:0006355    TAIR\n",
       "19     ath               AT1G01030                          AT1G01030               GO:0009908    TAIR"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "df_go[\"species\"] = \"ath\"\n",
    "df_go[\"unique_gene_identifiers\"] = df_go[\"locus\"]\n",
    "df_go[\"other_gene_identifiers\"] = \"\"\n",
    "df_go[\"gene_models\"] = df_go[\"unique_gene_identifiers\"].map(lambda x: \"\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_go[\"descriptions\"] = \"\"\n",
    "df_go[\"annotations\"] = df_go[\"term_id\"]\n",
    "df_go[\"sources\"] = \"TAIR\"\n",
    "high_confidence_categories = [\"experimental\",\"author_statement\",\"curator_statement\"]\n",
    "df_go[\"high_confidence\"] = df_go[\"evidence_code\"].apply(lambda x: EVIDENCE_CODES[x] in high_confidence_categories)\n",
    "\n",
    "# Subset to only include the high-quality GO annotations and ouptut the dataset to a csv file.\n",
    "df_go_high_confidence = df_go[df_go[\"high_confidence\"]==True]\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_curated_go_annotations.csv\")\n",
    "df_go_high_confidence = df_go_high_confidence[reshaped_columns]\n",
    "df_go_high_confidence.to_csv(path, index=False)\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "df_go = df_go[reshaped_columns]\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_all_go_annotations.csv\")\n",
    "df_go.to_csv(path, index=False)\n",
    "df_go.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### File with plant ontology term annotations (po_[termporal/anatomy]_gene_arabidopsis_tair.assoc)\n",
    "There are two separate files available that include annotations of PO terms. The files do not have headers so column names are added based on how the columns are described in the accompanying available readme files. One of the files contains annotations for PO terms that are spatial, or describe a specific part of plant anatomy or plant molecular structures. The other file contains annotations for PO terms that are temporal, or refer to a specific process or stage of development. These files are each read in separately, and the next cells look at the quantity of unique values in the columns of each dataset. There are more spatial annotations than temporal annotations, and a greater number of terms used to describe the spatial annotations.\n",
    "\n",
    "The next field combines the two datasets of PO annotations and looks at the number of unique values for each column in the resulting dataset. Because there is no overlap in the terms between the two, the datasets are simply appended to one another and the total unique terms are a sum of the individual datasets.\n",
    "\n",
    "Each term annotation in this dataset is also associated with an evidence code specifying the method by which this annotation was made, which is related to the confidence that we can have in this annotation, and the tasks that the annotation should be used for. Almost all of the PO term annotations are high confidence, they are experimentally validated, and only a few of them are derived from author statements.\n",
    "\n",
    "The strings which are described in the synonyms column are included as references to each gene, and are combined with the gene name mentioned in the symbol column into a single bar delimited list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(366398, 7)\n",
      "symbol               24123\n",
      "term_id                365\n",
      "references            2600\n",
      "evidence_code            6\n",
      "name                 25269\n",
      "synonyms             28005\n",
      "type                     6\n"
     ]
    }
   ],
   "source": [
    "# Reading in the dataset of spatial PO term annotations.\n",
    "filename = \"../databases/tair/po_anatomy_gene_arabidopsis_tair.assoc\"\n",
    "df_po_spatial = pd.read_table(filename, header=None, skiprows=1, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_spatial.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_spatial[col].values)) for col in df_po_spatial.columns}\n",
    "print(df_po_spatial.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(186740, 7)\n",
      "symbol               18015\n",
      "term_id                 84\n",
      "references             447\n",
      "evidence_code            4\n",
      "name                 18506\n",
      "synonyms             18960\n",
      "type                     5\n"
     ]
    }
   ],
   "source": [
    "# Reading in the dataset of temporal PO term annotations.\n",
    "filename = \"../databases/tair/po_temporal_gene_arabidopsis_tair.assoc\"\n",
    "df_po_temporal = pd.read_table(filename, header=None, skiprows=1, usecols=[2,4,5,6,9,10,11])\n",
    "df_po_temporal.columns = [\"symbol\",\"term_id\",\"references\",\"evidence_code\",\"name\",\"synonyms\",\"type\"]\n",
    "unique_values = {col:len(pd.unique(df_po_temporal[col].values)) for col in df_po_temporal.columns}\n",
    "print(df_po_temporal.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   symbol                                           synonyms evidence_code\n",
      "0  ATARCA  AT1G18080|ATARCA|RACK1A_AT|RACK1A|SAC53|AtRACK...           IEP\n",
      "1    HMG1  AT1G76490|HMG1|HMGR1|AtHMGR1|MAD3|hydroxy meth...           TAS\n",
      "2    BAS1  AT2G26710|BAS1|CYP734A1|CYP72B1|PHYB ACTIVATIO...           TAS\n",
      "3    BAK1  AT4G33430|BAK1|RKS10|SERK3|ELG|ATSERK3|ATBAK1|...           TAS\n",
      "4    BRI1  AT4G39400|BRI1|CBB2|DWF2|BIN1|ATBRI1|BRASSINOS...           TAS\n",
      "5    ATJ6    AT5G06910|ATJ6|J-domain protein 6|MOJ9.8|MOJ9_8           IEP\n",
      "6    XPB2  AT5G41360|XPB2|ATXPB2|homolog of Xeroderma pig...           NAS\n",
      "7    XPB1  AT5G41370|XPB1|ATXPB1|homolog of xeroderma pig...           NAS\n",
      "8   PPCK2  AT3G04530|PPCK2|PEPCK2|ATPPCK2|phosphoenolpyru...           IEP\n",
      "9    MYC2  AT1G32640|ATMYC2|RD22BP1|JAI1|JIN1|MYC2|ZBF1|J...           NAS\n",
      "(553138, 7)\n",
      "symbol               24188\n",
      "term_id                449\n",
      "references            2671\n",
      "evidence_code            6\n",
      "name                 25350\n",
      "synonyms             28150\n",
      "type                     7\n"
     ]
    }
   ],
   "source": [
    "# Looking at how many unique values each column has.\n",
    "df_po = df_po_spatial.append(df_po_temporal, ignore_index=True)\n",
    "unique_values = {col:len(pd.unique(df_po[col].values)) for col in df_po.columns}\n",
    "print(df_po[[\"symbol\",\"synonyms\",\"evidence_code\"]].head(10))\n",
    "print(df_po.shape)\n",
    "for k,v in unique_values.items():\n",
    "    print(\"{:18}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "experimental               553020\n",
      "computational                   0\n",
      "phylogenetics                   0\n",
      "high_throughput                 0\n",
      "author_statement              118\n",
      "curator_statement               0\n",
      "electronic_annotation           0\n"
     ]
    }
   ],
   "source": [
    "# Quantifying the number of annotations of each type.\n",
    "code_quantities = {c:len([x for x in df_po[\"evidence_code\"] if EVIDENCE_CODES[x] in c]) \n",
    "             for c in list(set(EVIDENCE_CODES.values()))}\n",
    "for k,v in code_quantities.items():\n",
    "    print(\"{:25}{:8}\".format(k,v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol           object\n",
       "term_id          object\n",
       "references       object\n",
       "evidence_code    object\n",
       "name             object\n",
       "synonyms         object\n",
       "type             object\n",
       "dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_po['name'] = df_po['name'].astype(str)\n",
    "df_po.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>other_gene_identifiers</th>\n",
       "      <th>gene_models</th>\n",
       "      <th>descriptions</th>\n",
       "      <th>annotations</th>\n",
       "      <th>sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATARCA|AT1G18080</td>\n",
       "      <td>RACK1A_AT|RACK1A|SAC53|AtRACK1|RECEPTOR FOR AC...</td>\n",
       "      <td>AT1G18080</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ath</td>\n",
       "      <td>HMG1|AT1G76490</td>\n",
       "      <td>HMGR1|AtHMGR1|MAD3|hydroxy methylglutaryl CoA ...</td>\n",
       "      <td>AT1G76490</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ath</td>\n",
       "      <td>BAS1|AT2G26710</td>\n",
       "      <td>CYP734A1|CYP72B1|PHYB ACTIVATION TAGGED SUPPRE...</td>\n",
       "      <td>AT2G26710</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ath</td>\n",
       "      <td>BAK1|AT4G33430</td>\n",
       "      <td>RKS10|SERK3|ELG|ATSERK3|ATBAK1|BRI1-associated...</td>\n",
       "      <td>AT4G33430</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ath</td>\n",
       "      <td>BRI1|AT4G39400</td>\n",
       "      <td>CBB2|DWF2|BIN1|ATBRI1|BRASSINOSTEROID INSENSIT...</td>\n",
       "      <td>AT4G39400</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATJ6|AT5G06910</td>\n",
       "      <td>J-domain protein 6|MOJ9.8|MOJ9_8</td>\n",
       "      <td>AT5G06910</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>XPB2|AT5G41360</td>\n",
       "      <td>ATXPB2|homolog of Xeroderma pigmentosum comple...</td>\n",
       "      <td>AT5G41360</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ath</td>\n",
       "      <td>XPB1|AT5G41370</td>\n",
       "      <td>ATXPB1|homolog of xeroderma pigmentosum comple...</td>\n",
       "      <td>AT5G41370</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000003</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ath</td>\n",
       "      <td>PPCK2|AT3G04530</td>\n",
       "      <td>PEPCK2|ATPPCK2|phosphoenolpyruvate carboxylase...</td>\n",
       "      <td>AT3G04530</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000013</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ath</td>\n",
       "      <td>MYC2|AT1G32640</td>\n",
       "      <td>ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...</td>\n",
       "      <td>AT1G32640</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000017</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATSBT5.2|AT1G20160</td>\n",
       "      <td>CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...</td>\n",
       "      <td>AT1G20160</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>ATSBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td>AT1G32960</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.12|AT4G21326</td>\n",
       "      <td>ATSBT3.12|subtilase 3.12</td>\n",
       "      <td>AT4G21326</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12</td>\n",
       "      <td>AT5G51750</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000036</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ath</td>\n",
       "      <td>PWO1|AT3G03140</td>\n",
       "      <td>WWP-DOMAIN INTERACTOR OF POLYCOMBS11|T17B22.17...</td>\n",
       "      <td>AT3G03140</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000037</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12</td>\n",
       "      <td>AT5G51750</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000037</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ath</td>\n",
       "      <td>SEP3|AT1G24260</td>\n",
       "      <td>AGL9|SEPALLATA3|AGAMOUS-like 9|F3I6.19|F3I6_19...</td>\n",
       "      <td>AT1G24260</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000229</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ath</td>\n",
       "      <td>RABA1b|AT1G16920</td>\n",
       "      <td>RAB11|ATRABA1B|BEX5|RAB GTPase homolog A1B|BFA...</td>\n",
       "      <td>AT1G16920</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000256</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>MYC2|AT1G32640</td>\n",
       "      <td>ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...</td>\n",
       "      <td>AT1G32640</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000256</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>ATSBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td>AT1G32960</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ath</td>\n",
       "      <td>EXPA1|AT1G69530</td>\n",
       "      <td>ATEXPA1|EXP1|AT-EXP1|ATEXP1|ATHEXP ALPHA 1.2|e...</td>\n",
       "      <td>AT1G69530</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT1.3|AT5G51750</td>\n",
       "      <td>ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12</td>\n",
       "      <td>AT5G51750</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.12|AT5G59090</td>\n",
       "      <td>ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5</td>\n",
       "      <td>AT5G59090</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.13|AT5G59120</td>\n",
       "      <td>ATSBT4.13|subtilase 4.13|MNC17.1</td>\n",
       "      <td>AT5G59120</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT5.4|AT5G59810</td>\n",
       "      <td>ATSBT5.4|MMN10.6|MMN10_6</td>\n",
       "      <td>AT5G59810</td>\n",
       "      <td></td>\n",
       "      <td>PO:0000293</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ath</td>\n",
       "      <td>UBP19|AT2G24640</td>\n",
       "      <td>ubiquitin-specific protease 19|F25P17.6|F25P17_6</td>\n",
       "      <td>AT2G24640</td>\n",
       "      <td></td>\n",
       "      <td>PO:0002000</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ath</td>\n",
       "      <td>ATSBT5.2|AT1G20160</td>\n",
       "      <td>CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...</td>\n",
       "      <td>AT1G20160</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT3.3|AT1G32960</td>\n",
       "      <td>ATSBT3.3|F9L11.13|F9L11_13</td>\n",
       "      <td>AT1G32960</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT4.12|AT5G59090</td>\n",
       "      <td>ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5</td>\n",
       "      <td>AT5G59090</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ath</td>\n",
       "      <td>SBT5.4|AT5G59810</td>\n",
       "      <td>ATSBT5.4|MMN10.6|MMN10_6</td>\n",
       "      <td>AT5G59810</td>\n",
       "      <td></td>\n",
       "      <td>PO:0003011</td>\n",
       "      <td>TAIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species unique_gene_identifiers                             other_gene_identifiers gene_models descriptions annotations sources\n",
       "0      ath        ATARCA|AT1G18080  RACK1A_AT|RACK1A|SAC53|AtRACK1|RECEPTOR FOR AC...   AT1G18080               PO:0000003    TAIR\n",
       "1      ath          HMG1|AT1G76490  HMGR1|AtHMGR1|MAD3|hydroxy methylglutaryl CoA ...   AT1G76490               PO:0000003    TAIR\n",
       "2      ath          BAS1|AT2G26710  CYP734A1|CYP72B1|PHYB ACTIVATION TAGGED SUPPRE...   AT2G26710               PO:0000003    TAIR\n",
       "3      ath          BAK1|AT4G33430  RKS10|SERK3|ELG|ATSERK3|ATBAK1|BRI1-associated...   AT4G33430               PO:0000003    TAIR\n",
       "4      ath          BRI1|AT4G39400  CBB2|DWF2|BIN1|ATBRI1|BRASSINOSTEROID INSENSIT...   AT4G39400               PO:0000003    TAIR\n",
       "5      ath          ATJ6|AT5G06910                   J-domain protein 6|MOJ9.8|MOJ9_8   AT5G06910               PO:0000003    TAIR\n",
       "6      ath          XPB2|AT5G41360  ATXPB2|homolog of Xeroderma pigmentosum comple...   AT5G41360               PO:0000003    TAIR\n",
       "7      ath          XPB1|AT5G41370  ATXPB1|homolog of xeroderma pigmentosum comple...   AT5G41370               PO:0000003    TAIR\n",
       "8      ath         PPCK2|AT3G04530  PEPCK2|ATPPCK2|phosphoenolpyruvate carboxylase...   AT3G04530               PO:0000013    TAIR\n",
       "9      ath          MYC2|AT1G32640  ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...   AT1G32640               PO:0000017    TAIR\n",
       "10     ath      ATSBT5.2|AT1G20160  CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...   AT1G20160               PO:0000036    TAIR\n",
       "11     ath        SBT3.3|AT1G32960                         ATSBT3.3|F9L11.13|F9L11_13   AT1G32960               PO:0000036    TAIR\n",
       "12     ath       SBT3.12|AT4G21326                           ATSBT3.12|subtilase 3.12   AT4G21326               PO:0000036    TAIR\n",
       "13     ath        SBT1.3|AT5G51750           ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12   AT5G51750               PO:0000036    TAIR\n",
       "14     ath          PWO1|AT3G03140  WWP-DOMAIN INTERACTOR OF POLYCOMBS11|T17B22.17...   AT3G03140               PO:0000037    TAIR\n",
       "15     ath        SBT1.3|AT5G51750           ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12   AT5G51750               PO:0000037    TAIR\n",
       "16     ath          SEP3|AT1G24260  AGL9|SEPALLATA3|AGAMOUS-like 9|F3I6.19|F3I6_19...   AT1G24260               PO:0000229    TAIR\n",
       "17     ath        RABA1b|AT1G16920  RAB11|ATRABA1B|BEX5|RAB GTPase homolog A1B|BFA...   AT1G16920               PO:0000256    TAIR\n",
       "18     ath          MYC2|AT1G32640  ATMYC2|RD22BP1|JAI1|JIN1|ZBF1|JASMONATE INSENS...   AT1G32640               PO:0000256    TAIR\n",
       "19     ath        SBT3.3|AT1G32960                         ATSBT3.3|F9L11.13|F9L11_13   AT1G32960               PO:0000293    TAIR\n",
       "20     ath         EXPA1|AT1G69530  ATEXPA1|EXP1|AT-EXP1|ATEXP1|ATHEXP ALPHA 1.2|e...   AT1G69530               PO:0000293    TAIR\n",
       "21     ath        SBT1.3|AT5G51750           ATSBT1.3|subtilase 1.3|MIO24.12|MIO24_12   AT5G51750               PO:0000293    TAIR\n",
       "22     ath       SBT4.12|AT5G59090         ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5   AT5G59090               PO:0000293    TAIR\n",
       "23     ath       SBT4.13|AT5G59120                   ATSBT4.13|subtilase 4.13|MNC17.1   AT5G59120               PO:0000293    TAIR\n",
       "24     ath        SBT5.4|AT5G59810                           ATSBT5.4|MMN10.6|MMN10_6   AT5G59810               PO:0000293    TAIR\n",
       "25     ath         UBP19|AT2G24640   ubiquitin-specific protease 19|F25P17.6|F25P17_6   AT2G24640               PO:0002000    TAIR\n",
       "26     ath      ATSBT5.2|AT1G20160  CRSP|SBT5.2(a)|SBT5.2(b)|CO2 RESPONSE SECRETED...   AT1G20160               PO:0003011    TAIR\n",
       "27     ath        SBT3.3|AT1G32960                         ATSBT3.3|F9L11.13|F9L11_13   AT1G32960               PO:0003011    TAIR\n",
       "28     ath       SBT4.12|AT5G59090         ATSBT4.12|subtilase 4.12|K18B18.5|K18B18_5   AT5G59090               PO:0003011    TAIR\n",
       "29     ath        SBT5.4|AT5G59810                           ATSBT5.4|MMN10.6|MMN10_6   AT5G59810               PO:0003011    TAIR"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Restructuring the dataset to include all the expected column names.\n",
    "combine_columns = lambda row, columns: concatenate_with_delim(\"|\", [row[column] for column in columns])\n",
    "df_po[\"species\"] = \"ath\"\n",
    "df_po[\"unique_gene_identifiers\"] = df_po.apply(lambda x: combine_columns(x, [\"symbol\", \"name\"]), axis=1)\n",
    "df_po[\"other_gene_identifiers\"] = df_po[\"synonyms\"]\n",
    "df_po[\"gene_model_strings_1\"] =  df_po[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_po[\"gene_model_strings_2\"] =  df_po[\"unique_gene_identifiers\"].map(lambda x: \"|\".join([s for s in x.split(\"|\") if is_gene_model(s)]))\n",
    "df_po[\"gene_models\"] = df_po.apply(lambda x: combine_columns(x, [\"gene_model_strings_1\", \"gene_model_strings_2\"]), axis=1)\n",
    "df_po[\"descriptions\"] = \"\"\n",
    "df_po[\"annotations\"] = df_po[\"term_id\"]\n",
    "df_po[\"sources\"] = \"TAIR\"\n",
    "df_po[\"other_gene_identifiers\"] = df_po.apply(lambda row: subtract_string_lists(\"|\", row[\"other_gene_identifiers\"],row[\"unique_gene_identifiers\"]), axis=1)\n",
    "df_po = df_po[reshaped_columns]\n",
    "\n",
    "# Outputting the dataset of annotations to a csv file.\n",
    "path = os.path.join(OUTPUT_DIR,\"tair_curated_po_annotations.csv\")\n",
    "df_po.to_csv(path, index=False)\n",
    "df_po.head(30)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
